### PR TITLE
Refactored download_models.py

### PR DESCRIPTION
**Pull request by mshabunin**
:exclamation: This PR supersedes #1127  and #1126
- split functionality
- support multiple URLs
- reduce output
- updated google drive download mechanism
- support new yolo models

### CODE REVIEW

File: testdata/dnn/download_models.py [hunk 1]
```diff
@@ -5,215 +5,235 @@
...
+        rcode = response.status_code
+        rsize = int(response.headers.get('content-length', 0)) / self.MB
+        print('    {} [{:.2f} Mb]'.format(rcode, rsize))
 
-    def __str__(self):
-        return 'Model <{}>'.format(self.name)
+    def make_response(self, url, session):
+        pieces = urlparse(url)
+        if pieces.netloc in ["docs.google.com", "drive.google.com"]:
+            return session.get(url, params={'confirm': True}, stream=True, timeout=self.TIMEOUT)
```
Comment (mshabunin) [inline testdata/dnn/download_models.py:33]: For some reason this simple thing helps to download large files from Google Drive (Ubuntu 22, packaged requests). I'd appreciate additional checks performed by other people using different OSes and locations.

File: testdata/dnn/download_models.py [hunk 2]
```diff
@@ -5,215 +5,235 @@
...
-                    print('  ref - hash mismatch, removing')
-                    self.filename.unlink()
+            filename.unlink()
         except Exception as e:
-            print('  ref {}'.format(e))
+            print("  cleanup failed: {}".format(e))
 
-    def get_sub(self, ref, arch):
-        print('** {}'.format(self.filename))
+    def handle_bad_download(self, filename):
```
Comment (mshabunin) [inline testdata/dnn/download_models.py:164]: This function has been reduced, I hope the logic is similar to the original. Maybe some cornercases are not handled well, but in general I think it is OK.

File: testdata/dnn/download_models.py [hunk 3]
```diff
@@ -612,10 +662,13 @@ def handle_bad_download(self):
                 sha='6d45d2f06150e9045631c7928093728b07c8b12d',
                 filename='onnx/data/output_resnet50_int8.pb'),
         ]),
-
+    # TODO: bad file
```
Comment (mshabunin) [inline testdata/dnn/download_models.py:665]: Archive downloaded from GitHub has different checksum and contains different files, perhaps this model and related test should be reviewed separately.

File: testdata/dnn/download_models.py [hunk 4]
```diff
@@ -859,7 +912,10 @@ def handle_bad_download(self):
         filename='yolov4x-mish.weights'),
     Model(
         name='GSOC2016-GOTURN',  # https://github.com/opencv/opencv_contrib/issues/941
-        gdrive='1j4UTqVE4EGaUFiK7a5I_CYX7twO9c5br',
```
Comment (mshabunin) [inline testdata/dnn/download_models.py:862]: We detect Google Drive by URL now.

File: testdata/dnn/download_models.py [hunk 5]
```diff
@@ -933,9 +989,21 @@ def handle_bad_download(self):
...
         name='yolov7_not_simplified',
-        gdrive='1rm3mIqjJNu0xPTCjMKnXccspazV1B2zv',
+        url=[
+            'https://docs.google.com/uc?export=download&id=1ljSh81ydO5ylsnDoV_mt3zj5RX_cgLu8',
+            'https://dl.opencv.org/models/yolov7/yolov7_not_simplified.onnx'
+        ],
         sha='fcd0fa401c83bf2b29e18239a9c2c989c9b8669d',
         filename='onnx/models/yolov7_not_simplified.onnx'),
+    Model(
+        name='yolox_s_inf_decoder',
```
Comment (mshabunin) [inline testdata/dnn/download_models.py:999]: New model from another PR.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**